### PR TITLE
feat: support function updater just like useState

### PR DESCRIPTION
**Pull request by leftstick**
accept function updater as argument for setxxx.

For example: 

```javascript
const [value, setValue] = useLocalStorageState('cascader');

setValue(previousValue => [...previousValue, {value: 1, label: 'hello'}])
```

### CODE REVIEW

File: src/useStorageState/index.ts [hunk 1]
```diff
@@ -11,10 +11,16 @@ function useStorageState<T>(storage: Storage, key: string, defaultValue?: T | ((
...
     return defaultValue;
   });
-  function updateState(value?: T) {
+  function updateState(value?: T | ((previousState: T) => T)) {
     if (typeof value === 'undefined') {
       storage.removeItem(key);
       setState(defaultValue);
+    } else if (typeof value === 'function') {
+      const raw = storage.getItem(key);
+      const previousState = raw !== null ? JSON.parse(raw) : defaultValue;
```
Comment (awmleer) [inline src/useStorageState/index.ts]: 这两行的逻辑和 4-12 行的逻辑是相同的，所以可以考虑抽成一个函数出来？

此外，这里的 `defaultValue` 也应该向上面一样判断 `typeof defaultValue === 'function'` 的情况。。

Comment (leftstick) [reply src/useStorageState/index.ts]: 同意，我改一下你再看看

Comment (leftstick) [reply src/useStorageState/index.ts]: 改好了，来吧

File: src/useStorageState/index.ts [hunk 2]
```diff
@@ -1,20 +1,44 @@
...
+    return JSON.parse(raw);
+  }
+  if (isFunction<IInitFunc<T>>(defaultValue)) {
+    return defaultValue();
+  }
+  return defaultValue;
+}
+
+function useStorageState<T>(storage: Storage, key: string, defaultValue?: T | IInitFunc<T>) {
+  const [state, setState] = useState<T | undefined>(() => getByKey(storage, key, defaultValue));
```
Comment (awmleer) [inline src/useStorageState/index.ts]: 感觉做成闭包函数更简单也更清晰一些。。没必要多些那么多手动推导的类型。。

```suggestion
  const [state, setState] = useState<T | undefined>(() => getByKey(storage, key, defaultValue));
  function getByKey() {
    const raw = storage.getItem(key);
    if (raw) {
      return JSON.parse(raw);
    }
    if (typeof defaultValue === 'function') {
      return (defaultValue as () => T)();
    }
    return defaultValue;
  }
```

**Review by awmleer — CHANGES_REQUESTED** [review]

File: src/useStorageState/index.ts (hunk 2 above)
Comment (leftstick) [reply src/useStorageState/index.ts]: 思路好，我回去再改改

Comment (leftstick) [reply src/useStorageState/index.ts]: 改好了

**Review by brickspert — CHANGES_REQUESTED** [review]
Maybe you should change the `useSessionStorageValue` together~

File: src/useLocalStorageState/demo/demo3.jsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+export default function () {
+  const [user, setUser] = useLocalStorageState('user', {
+    id: 9234634791,
+    name: 'Zhangsan',
+    age: 33,
+  });
+  return (
+    <>
+      <Input
```
Comment (brickspert) [inline src/useLocalStorageState/demo/demo3.jsx:13]: this `Input` should have a width 200px

image

File: src/useLocalStorageState/index_cn.mdx [hunk 4]
```diff
@@ -30,13 +34,17 @@ import Demo2CodeJsx from '!raw-loader!./demo/demo2.jsx';
   <Demo2 />
 </JackBox>
 
+<JackBox jsCode={Demo3CodeJsx} tsCode={Demo3CodeTsx} demoName='使用function updater存储' description='useLocalStorageState 里也可以用function updater，就像useState那样.'>
```
Comment (brickspert) [inline src/useLocalStorageState/index_cn.mdx]: Space is required between Chinese and English

**Review by brickspert — CHANGES_REQUESTED** [review]

File: src/useLocalStorageState/demo/demo3.jsx (hunk 3 above)
Comment (leftstick) [reply src/useLocalStorageState/demo/demo3.jsx:13]: done

File: src/useLocalStorageState/index_cn.mdx (hunk 4 above)
Comment (leftstick) [reply src/useLocalStorageState/index_cn.mdx]: done

**Comment by leftstick** [comment]
> Maybe you should change the `useSessionStorageValue` together~

@brickspert  I will create new PR for that once this one approved, thanks for reminding.

**Review by brickspert — APPROVED** [review]
LGTM

**Review by awmleer — APPROVED** [review]
LGTM

**Pull request merged**